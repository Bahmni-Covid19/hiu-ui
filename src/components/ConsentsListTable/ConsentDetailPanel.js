import React from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import { Grid, Table, TableBody, TableCell, TableHead, TableRow } from '@material-ui/core';
import { formatDateString } from '../common/HealthInfo/FhirResourcesUtils';

const useStyles = makeStyles(theme => ({
    detailPanel: {
        paddingTop: theme.spacing(2),
        paddingBottom: theme.spacing(2),
        paddingLeft: theme.spacing(5),
        backgroundColor: theme.palette.background.default
    }
}));

const getHiTypesText = hiTypes => {
    return hiTypes.join(', ');
};

const getPermissionDateRangeText = permission => {
    let fromDate = formatDateString(permission.dateRange.from, false);
    let toDate = formatDateString(permission.dateRange.to, false);
    return `${fromDate} - ${toDate}`;
};

const consentExpiredComponent = (expiryTimestamp) => {
    return (
        <Typography variant="body2">
            Consent Request <strong>Expired</strong> on {formatDateString(expiryTimestamp, true)}
        </Typography>
    )
}

const consentDeniedComponent = (deniedTimestamp) => {
    return (
        <Typography variant="body2">
            Consent Request <strong>Denied</strong> on {formatDateString(deniedTimestamp, true)}
        </Typography>
    )
}

const consentArtefactDetailsComponent = (consentArtefacts) => {
    return consentArtefacts.length === 0 ?
        <Grid item xs={12}>
            <Typography variant="body2">
                <strong>No Consent Artefacts available</strong>
            </Typography>
        </Grid> :
        (
            <Grid container>
                <Grid item xs={6}><Typography variant="body2">
                    <strong>Granted HI Types:</strong> {getHiTypesText(consentArtefacts[0].hiTypes)}
                </Typography>
                </Grid>
                <Grid item xs={6}>
                    <Typography variant="body2">
                        <strong>Granted Date Range:</strong> {getPermissionDateRangeText(consentArtefacts[0].permission)}
                    </Typography>
                </Grid>
                <Grid item xs={12}>
                    <Typography variant="body2">
                        <strong>Consent Artefacts:</strong>
                    </Typography>
                    <Table>
                        <TableHead>
                            <TableRow>
                                <TableCell>Facility Name</TableCell>
                                <TableCell>Status</TableCell>
                                <TableCell>Status Last Updated on</TableCell>
                            </TableRow>
                        </TableHead>
                        <TableBody>
                            {consentArtefacts.map((consentArtefact, index) => {
                                return (<TableRow>
                                    <TableCell>{consentArtefact.hipName}</TableCell>
                                    <TableCell>{consentArtefact.status}</TableCell>
                                    <TableCell>{formatDateString(consentArtefact.dateModified, true)}</TableCell>
                                </TableRow>)
                            })}
                        </TableBody>
                    </Table>
                </Grid>
            </Grid>
        )
}


const ConsentDetailPanel = ({ consentDetail }) => {
    const classes = useStyles();
    const consentStatus = consentDetail.status;
    return (
        <Grid container spacing={1} className={classes.detailPanel}>
            <Grid item xs={6}>
                <Typography variant="body2">
                    <strong>Requested HI Types:</strong> {getHiTypesText(consentDetail.hiTypes)}
                </Typography>
            </Grid>
            <Grid item xs={6}>
                <Typography variant="body2">
                    <strong>Requested Date Range:</strong> {getPermissionDateRangeText(consentDetail.permission)}
                </Typography>
            </Grid>
            <Grid item xs={12}>
                {consentStatus === 'DENIED' && consentDeniedComponent(consentDetail.dateModified)}
                {consentStatus === 'EXPIRED' && consentExpiredComponent(consentDetail.expiredDate)}
                {(consentStatus === 'GRANTED' || consentStatus === 'REVOKED') && (
                    <Grid container>
                        {consentDetail.consentArtefacts.length === 0 ?
                            <Grid item xs={12}>
                                <Typography variant="body2">
                                    <strong>No Consent Artefacts available</strong>
                                </Typography>
                            </Grid>
                            :
                            (
                                consentArtefactDetailsComponent(consentDetail.consentArtefacts)
                            )}
                    </Grid>
                )}
            </Grid>
        </Grid>
    );
};

ConsentDetailPanel.propTypes = {
    consentDetail: PropTypes.object.isRequired
};

export default ConsentDetailPanel;
