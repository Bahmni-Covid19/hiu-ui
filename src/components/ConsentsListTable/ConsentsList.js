const ConsentsList = [
  {
    status: "REQUESTED",
    createdDate: "2020-01-27T07:08:48.843+0000",
    expiredDate: "2020-03-12T07:08:00.000+0000",
    purpose: {
      text: "REMOTE_CONSULTING",
      code: "RemoteConsulting",
      refUri: null
    },
    patient: {
      id: "vishwanathan.iyer@ncg",
      firstName: "Vishwanathan",
      lastName: "Iyer"
    },
    permission: {
      accessMode: "VIEW",
      dateRange: {
        from: "2020-01-01T07:08:00.000+0000",
        to: "2020-01-08T07:08:00.000+0000"
      },
      frequency: {
        unit: "HOUR",
        value: 1,
        repeats: 0
      }
    },
    id: "5c71c78e-a77e-4dd5-9f0f-6c3095118317"
  },
  {
    status: "REQUESTED",
    createdDate: "2020-01-27T07:08:48.843+0000",
    expiredDate: "2020-03-12T07:08:00.000+0000",
    purpose: {
      text: "REMOTE_CONSULTING",
      code: "RemoteConsulting",
      refUri: null
    },
    patient: {
      id: "vishwanathan.iyer@ncg",
      firstName: "Vishwanathan",
      lastName: "Iyer"
    },
    permission: {
      accessMode: "VIEW",
      dateRange: {
        from: "2020-01-01T07:08:00.000+0000",
        to: "2020-01-08T07:08:00.000+0000"
      },
      frequency: {
        unit: "HOUR",
        value: 1,
        repeats: 0
      }
    },
    id: "5c71c78e-a77e-4dd5-9f0f-6c3095118327"
  },
  {
    status: "GRANTED",
    createdDate: "2020-01-27T07:08:48.843+0000",
    expiredDate: "2020-03-12T07:08:00.000+0000",
    approvedDate: "2020-01-27T07:08:48.843+0000",
    purpose: {
      text: "REMOTE_CONSULTING",
      code: "RemoteConsulting",
      refUri: null
    },
    patient: {
      id: "vishwanathan.iyer@ncg",
      firstName: "Vishwanathan",
      lastName: "Iyer"
    },
    permission: {
      accessMode: "VIEW",
      dateRange: {
        from: "2020-01-01T07:08:00.000+0000",
        to: "2020-01-08T07:08:00.000+0000"
      },
      frequency: {
        unit: "HOUR",
        value: 1,
        repeats: 0
      }
    },
    id: "5c71c78e-a77e-4dd5-9f0f-6c3095118337"
  },
  {
    status: "REQUESTED",
    createdDate: "2020-01-27T07:08:48.843+0000",
    expiredDate: "2020-03-12T07:08:00.000+0000",
    purpose: {
      text: "REMOTE_CONSULTING",
      code: "RemoteConsulting",
      refUri: null
    },
    patient: {
      id: "vishwanathan.iyer@ncg",
      firstName: "Vishwanathan",
      lastName: "Iyer"
    },
    permission: {
      accessMode: "VIEW",
      dateRange: {
        from: "2020-01-01T07:08:00.000+0000",
        to: "2020-01-08T07:08:00.000+0000"
      },
      frequency: {
        unit: "HOUR",
        value: 1,
        repeats: 0
      }
    },
    id: "5c71c78e-a77e-4dd5-9f0f-6c3095118347"
  },
  {
    status: "GRANTED",
    createdDate: "2020-01-27T07:08:48.843+0000",
    expiredDate: "2020-03-12T07:08:00.000+0000",
    approvedDate: "2020-01-27T07:08:48.843+0000",
    purpose: {
      text: "REMOTE_CONSULTING",
      code: "RemoteConsulting",
      refUri: null
    },
    patient: {
      id: "vishwanathan.iyer@ncg",
      firstName: "Vishwanathan",
      lastName: "Iyer"
    },
    permission: {
      accessMode: "VIEW",
      dateRange: {
        from: "2020-01-01T07:08:00.000+0000",
        to: "2020-01-08T07:08:00.000+0000"
      },
      frequency: {
        unit: "HOUR",
        value: 1,
        repeats: 0
      }
    },
    id: "5c71c78e-a77e-4dd5-9f0f-6c30951183257"
  },
  {
    status: "GRANTED",
    createdDate: "2020-01-27T07:08:48.843+0000",
    expiredDate: "2020-03-12T07:08:00.000+0000",
    approvedDate: "2020-01-27T07:08:48.843+0000",
    purpose: {
      text: "REMOTE_CONSULTING",
      code: "RemoteConsulting",
      refUri: null
    },
    patient: {
      id: "vishwanathan.iyer@ncg",
      firstName: "Vishwanathan",
      lastName: "Iyer"
    },
    permission: {
      accessMode: "VIEW",
      dateRange: {
        from: "2020-01-01T07:08:00.000+0000",
        to: "2020-01-08T07:08:00.000+0000"
      },
      frequency: {
        unit: "HOUR",
        value: 1,
        repeats: 0
      }
    },
    id: "5c71c78e-a77e-4dd5-9f0f-6c3095118367"
  },
  {
    status: "REQUESTED",
    createdDate: "2020-01-27T07:08:48.843+0000",
    expiredDate: "2020-03-12T07:08:00.000+0000",
    purpose: {
      text: "REMOTE_CONSULTING",
      code: "RemoteConsulting",
      refUri: null
    },
    patient: {
      id: "vishwanathan.iyer@ncg",
      firstName: "Vishwanathan",
      lastName: "Iyer"
    },
    permission: {
      accessMode: "VIEW",
      dateRange: {
        from: "2020-01-01T07:08:00.000+0000",
        to: "2020-01-08T07:08:00.000+0000"
      },
      frequency: {
        unit: "HOUR",
        value: 1,
        repeats: 0
      }
    },
    id: "5c71c78e-a77e-4dd5-9f0f-6c3095118377"
  },
  {
    status: "REQUESTED",
    createdDate: "2020-01-27T07:08:48.843+0000",
    expiredDate: "2020-03-12T07:08:00.000+0000",
    purpose: {
      text: "REMOTE_CONSULTING",
      code: "RemoteConsulting",
      refUri: null
    },
    patient: {
      id: "vishwanathan.iyer@ncg",
      firstName: "Vishwanathan",
      lastName: "Iyer"
    },
    permission: {
      accessMode: "VIEW",
      dateRange: {
        from: "2020-01-01T07:08:00.000+0000",
        to: "2020-01-08T07:08:00.000+0000"
      },
      frequency: {
        unit: "HOUR",
        value: 1,
        repeats: 0
      }
    },
    id: "5c71c78e-a77e-4dd5-9f0f-6c3095118387"
  }
];

export default ConsentsList;
