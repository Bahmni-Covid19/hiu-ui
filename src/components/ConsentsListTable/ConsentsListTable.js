import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import * as PropTypes from 'prop-types';
import ArrowForwardIosIcon from '@material-ui/icons/ArrowForwardIos';
import MaterialTable from 'material-table';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import { formatDateString } from '../common/HealthInfo/FhirResourcesUtils';

const useStyles = makeStyles((theme) => ({
  table: {
    clear: 'both',
    backgroundColor: theme.palette.primary.main,
  },
  title: {
    textTransform: 'uppercase',
  },
}));

const ConsentsListTable = ({ loadConsents, consentsList }) => {
  const classes = useStyles();
  const [refreshCounter, setRefreshCounter] = useState(0);
  useEffect(() => {
    loadConsents();
  }, [refreshCounter]);

  const headerRow = {
    name: 'Name',
    jataayuId: 'Jataayu ID',
    requestStatus: 'Request Status',
    consentGrantedDate: 'Consent granted on',
    consentExpiryDate: 'Consent expiry on',
    consentCreatedDate: 'Consent created on',
  };

  function isGrantedConsent(status) {
    return status === 'GRANTED';
  }

  function getPatientFullName(patient) {
    return `${patient.firstName} ${patient.lastName}`;
  }

  return (
    <div style={{ maxWidth: '100%' }}>
      <MaterialTable
        options={{
          headerStyle: {
            backgroundColor: '#009688',
            color: '#FFF',
          },
        }}
        className={classes.table}
        columns={[
          { title: headerRow.name, field: 'name' },
          { title: headerRow.jataayuId, field: 'id' },
          { title: headerRow.requestStatus, field: 'status' },
          { title: headerRow.consentGrantedDate, field: 'grantedOn' },
          { title: headerRow.consentExpiryDate, field: 'expiredOn' },
          { title: headerRow.consentCreatedDate, field: 'createdOn' },
          { title: '', field: 'navLink', width: 50 },
        ]}
        data={consentsList.map((consent) => ({
          name: getPatientFullName(consent.patient),
          id: consent.patient.id,
          status: isGrantedConsent(consent.status)
            ? 'Consent granted'
            : 'Request sent',
          grantedOn: isGrantedConsent(consent.status)
            ? formatDateString(consent.approvedDate)
            : '-',
          expiredOn: isGrantedConsent(consent.status)
            ? formatDateString(consent.expiredDate)
            : '-',
          createdOn: formatDateString(consent.createdDate),
          navLink: isGrantedConsent(consent.status) ? (
            <Link to={`/health-info/${consent.id}`}>
              <ArrowForwardIosIcon color="primary" />
            </Link>
          ) : (
            ''
          ),
        }))}
        title={(
          <Typography className={classes.title} variant="h5">
            Consent List
          </Typography>
        )}
        actions={[
          {
            icon: 'refresh',
            tooltip: 'Refresh',
            isFreeAction: true,
            onClick: () => setRefreshCounter(refreshCounter + 1)
          }
        ]}
      />
    </div>
  );
};

const consentShape = PropTypes.shape({
  id: PropTypes.string,
  patient: PropTypes.shape({
    id: PropTypes.string,
    firstName: PropTypes.string,
    lastName: PropTypes.string,
  }),
  status: PropTypes.string,
  approvedDate: PropTypes.string,
  expiredDate: PropTypes.string,
  createdDate: PropTypes.string,
});

ConsentsListTable.propTypes = {
  loadConsents: PropTypes.func.isRequired,
  consentsList: PropTypes.arrayOf(consentShape).isRequired,
};

export default ConsentsListTable;
