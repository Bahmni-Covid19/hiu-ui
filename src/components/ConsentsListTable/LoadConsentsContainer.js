import React from "react";
import { connect } from "react-redux";
import LoadConsents from "./ConsentsList";
import { loadConsents } from "../../redux/actions/loadConsentsActions";
import getNestedObject from "../../utils/getNestedObject";

const mapStateToProps = state => ({
  huiId: "1",
  success: state.loadConsents.success,
  loading: state.loadConsents.loading,
  error: state.loadConsents.error,
  serverError: state.loadConsents.serverError
});

const mapDispatchToProps = {
  loadConsents
};

export default connect(mapStateToProps, mapDispatchToProps)(LoadConsents);
