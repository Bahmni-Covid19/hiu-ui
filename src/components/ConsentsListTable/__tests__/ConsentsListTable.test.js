import React from 'react';
import { shallow } from 'enzyme';
import MaterialTable from 'material-table';
import ConsentsListTable from '../ConsentsListTable';

describe('ConsentsListTable', () => {
  const loadConsents = jest.fn();
  const mockData = [
    {
      patient: {
        firstName: 'John',
        lastName: 'Smith',
      },
      status: 'GRANTED',
      approvedDate: '2020-02-19T00:30:08.000+0000',
      expiredDate: '2020-02-20T03:17:48.603+0000',
      navLink: '',
    },
  ];
  const wrapper = shallow(
    <ConsentsListTable consentsList={mockData} loadConsents={loadConsents} theme="" />,
  );

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
    // expect(loadConsents).toHaveBeenCalled();
  });
  it('should render data in consent list when status is granted', () => {
    expect(wrapper.find(MaterialTable).length).toEqual(1);
    expect(wrapper.find(MaterialTable).props().data[0].name).toEqual('John Smith');
    expect(wrapper.find(MaterialTable).props().data[0].status).toEqual('Consent granted');
    expect(wrapper.find(MaterialTable).props().data[0].grantedOn).toEqual('19/2/2020');
    expect(wrapper.find(MaterialTable).props().data[0].expiredOn).toEqual('20/2/2020');
    expect(wrapper.find(MaterialTable).props().data[0].navLink).not.toEqual('');
  });

  it('should render data in consent list when status is not granted', () => {
    mockData[0].status = 'DENIED';
    wrapper.setProps({ consentsList: mockData });

    expect(wrapper.find(MaterialTable).length).toEqual(1);
    expect(wrapper.find(MaterialTable).props().data[0].name).toEqual('John Smith');
    expect(wrapper.find(MaterialTable).props().data[0].status).toEqual('Request sent');
    expect(wrapper.find(MaterialTable).props().data[0].grantedOn).toEqual('-');
    expect(wrapper.find(MaterialTable).props().data[0].expiredOn).toEqual('-');
    expect(wrapper.find(MaterialTable).props().data[0].navLink).toEqual('');
  });
});
