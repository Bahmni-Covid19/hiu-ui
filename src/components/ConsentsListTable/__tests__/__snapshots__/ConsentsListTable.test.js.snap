// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsentsListTable should render properly 1`] = `
<div
  style={
    Object {
      "maxWidth": "100%",
    }
  }
>
  <WithStyles(Component)
    actions={
      Array [
        Object {
          "icon": "refresh",
          "isFreeAction": true,
          "onClick": [Function],
          "tooltip": "Refresh",
        },
      ]
    }
    className="makeStyles-table-1"
    columns={
      Array [
        Object {
          "field": "name",
          "title": "Name",
        },
        Object {
          "field": "id",
          "title": "Jataayu ID",
        },
        Object {
          "field": "status",
          "title": "Request Status",
        },
        Object {
          "customSort": [Function],
          "field": "createdOn",
          "title": "Consent created on",
        },
        Object {
          "customSort": [Function],
          "field": "grantedOn",
          "title": "Consent granted on",
        },
        Object {
          "customSort": [Function],
          "field": "expiredOn",
          "title": "Consent expiry on",
        },
        Object {
          "field": "navLink",
          "title": "",
          "width": 50,
        },
      ]
    }
    data={
      Array [
        Object {
          "createdOn": "18/02/2020 08:47",
          "expiredOn": "20/02/2020 08:47",
          "grantedOn": "19/02/2020 06:00",
          "id": undefined,
          "name": "John Smith",
          "navLink": <ForwardRef
            to="/health-info/undefined"
          >
            <UNDEFINED
              color="primary"
            />
          </ForwardRef>,
          "status": "Consent granted",
        },
      ]
    }
    isLoading={false}
    options={
      Object {
        "headerStyle": Object {
          "backgroundColor": "#009688",
          "color": "#FFF",
        },
      }
    }
    title={
      <ForwardRef(WithStyles)
        className="makeStyles-title-2"
        variant="h5"
      >
        Consent List
      </ForwardRef(WithStyles)>
    }
  />
</div>
`;
